<a href="https://colab.research.google.com/github/Sayanik-tech/Deep_Learning-ANN/blob/main/ANN_Churn_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf

In [16]:
tf.__version__

'2.6.0'

**Loading Data set**

In [4]:
dataset = pd.read_csv('Churn_Modelling.csv')
dataset.head()

RowNumber  CustomerId   Surname  ...  IsActiveMember EstimatedSalary Exited
0          1    15634602  Hargrave  ...               1       101348.88      1
1          2    15647311      Hill  ...               1       112542.58      0
2          3    15619304      Onio  ...               0       113931.57      1
3          4    15701354      Boni  ...               0        93826.63      0
4          5    15737888  Mitchell  ...               1        79084.10      0

[5 rows x 14 columns]

In [5]:
x = dataset.iloc[: ,3:-1].values
y = dataset.iloc[: , -1].values

In [6]:
print(x)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


In [7]:
print(y)

[1 0 1 ... 1 1 0]


**Encoding Categorical data**

**Label Encoding The 'Gender' column**

In [8]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
x[: , 2] = le.fit_transform(x[: , 2])

In [9]:
print(x)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


**One Hot Encoding The 'Geography' column**

In [10]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(),[1])],remainder = 'passthrough')
x=np.array(ct.fit_transform(x))
print(x)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


**Train,Test Splitting**

In [11]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2,random_state = 1)

**Feature Scaling**

In [12]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)
print(x_train)

[[-0.99850112  1.71490137 -0.57273139 ... -1.55337352  0.97725852
   0.42739449]
 [ 1.00150113 -0.58312392 -0.57273139 ... -1.55337352 -1.02327069
  -1.02548708]
 [-0.99850112  1.71490137 -0.57273139 ...  0.64376017  0.97725852
  -0.94479772]
 ...
 [ 1.00150113 -0.58312392 -0.57273139 ...  0.64376017  0.97725852
  -0.14096853]
 [ 1.00150113 -0.58312392 -0.57273139 ...  0.64376017  0.97725852
   0.01781218]
 [-0.99850112  1.71490137 -0.57273139 ...  0.64376017 -1.02327069
  -1.15822478]]


**Initializing the ANN**

In [23]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
ann = Sequential()

**Adding Input Layer and Hidden Layer**

In [27]:
ann.add(Dense(units= 16, activation='relu',kernel_initializer='he_uniform'))

In [28]:
ann.add(Dense(units=8 , activation = 'sigmoid'))

In [29]:
ann.add(Dense(units=4 , activation = 'relu',kernel_initializer='he_uniform'))

**Adding Output Layer**

In [31]:
ann.add(Dense(units=1 , activation = 'sigmoid'))

**Compilation**

In [33]:
ann.compile(optimizer= 'adam' , loss='binary_crossentropy' , metrics=['accuracy'])

**Training The ANN on Training Set**

In [34]:
ann.fit(x_train,y_train,batch_size=32,epochs=100)

Epoch 1/100
250/250 [==============================] - 4s 2ms/step - loss: 0.6636 - accuracy: 0.7222
Epoch 2/100
250/250 [==============================] - 0s 2ms/step - loss: 0.5872 - accuracy: 0.7972
Epoch 3/100
250/250 [==============================] - 0s 2ms/step - loss: 0.5165 - accuracy: 0.7972
Epoch 4/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4817 - accuracy: 0.7972
Epoch 5/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4658 - accuracy: 0.7972
Epoch 6/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4581 - accuracy: 0.7972
Epoch 7/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4532 - accuracy: 0.7972
Epoch 8/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4493 - accuracy: 0.7972
Epoch 9/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4462 - accuracy: 0.7972
Epoch 10/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4435 - accura

    -----Predicting the result of a single observation
Homework

Use our ANN model to predict if the customer with the following informations will leave the bank:

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: $ 60000

Number of Products: 2

Does this customer have a credit card ? Yes

Is this customer an Active Member: Yes

Estimated Salary: $ 50000

So, should we say goodbye to that customer ? **bold text**



In [35]:
ann.predict(sc.transform([[1,0,0,600,1,40,3,60000,2,1,1,50000]]))

array([[0.03338468]], dtype=float32)

**The Customer will not leave the bank**

Predicting The test Set Results

In [36]:
y_pred = ann.predict(x_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_test.reshape(len(y_test),1), y_pred.reshape(len(y_pred),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 1]]


Making The Confusion Matrix

In [37]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm = confusion_matrix(y_test,y_pred)
print(cm)

[[1497   88]
 [ 183  232]]


1497 correct predictions that customer stays with the bank (TP)//
232 correct predictions the customer leaves the bank (TN)//
88 incorrect predictions that customers leaves the bank (FN)//
183 incorrect predictions that customer stays with the bank (FP)//

In [38]:
accuracy_score(y_test,y_pred)

0.8645